File: examples/langchain_python/sql_agent.py
"""LangChain SQL agent for database queries."""

from langchain.agents import initialize_agent, AgentType
from langchain.tools import tool
from langchain_openai import ChatOpenAI


@tool
def query_database(query: str) -> str:
    """Execute a SQL query on the database."""
    # Mock implementation
    return "Query results: [{'id': 1, 'name': 'John'}, {'id': 2, 'name': 'Jane'}]"


@tool
def get_table_schema(table_name: str) -> str:
    """Get the schema of a database table."""
    # Mock implementation
    return f"Schema for {table_name}: id (INT), name (VARCHAR), email (VARCHAR)"


@tool
def list_tables() -> str:
    """List all tables in the database."""
    # Mock implementation
    return "Available tables: users, orders, products, customers"


@tool
def generate_sql(description: str) -> str:
    """Generate SQL query from natural language description."""
    # Mock implementation
    return f"SELECT * FROM users WHERE {description}"


# Initialize the language model
llm = ChatOpenAI(model="gpt-4", temperature=0)

# Create the tools list
tools = [query_database, get_table_schema, list_tables, generate_sql]

# Initialize the SQL agent
sql_agent = initialize_agent(
    tools=tools,
    llm=llm,
    agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION,
    agent_name="SQL Database Agent",
    verbose=True,
    max_iterations=5,
)

# Example usage
if __name__ == "__main__":
    response = sql_agent.run(
        "Find all active users who placed orders in the last 30 days"
    )
    print(response)

File: examples/langchain_typescript/sql_agent.ts
/**
 * LangChain TypeScript SQL agent for database queries
 */

import { AgentExecutor } from "langchain/agents";
import { ChatOpenAI } from "@langchain/openai";
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import { SqlDatabase } from "langchain/sql_db";

// Define SQL tools
const queryDatabase = tool({
  name: "query_database",
  description: "Execute a SQL query on the database",
  schema: z.object({
    query: z.string().describe("SQL query to execute"),
  }),
  func: async ({ query }) => {
    // Mock implementation
    return `Query results: [{"id": 1, "name": "John"}, {"id": 2, "name": "Jane"}]`;
  },
});

const getTableSchema = tool({
  name: "get_table_schema",
  description: "Get the schema of a database table",
  schema: z.object({
    tableName: z.string().describe("Name of the table"),
  }),
  func: async ({ tableName }) => {
    // Mock implementation
    return `Schema for ${tableName}: id (INT), name (VARCHAR), email (VARCHAR)`;
  },
});

const listTables = tool({
  name: "list_tables",
  description: "List all tables in the database",
  schema: z.object({}),
  func: async () => {
    // Mock implementation
    return "Available tables: users, orders, products, customers";
  },
});

const generateSql = tool({
  name: "generate_sql",
  description: "Generate SQL query from natural language description",
  schema: z.object({
    description: z.string().describe("Natural language description of the query"),
  }),
  func: async ({ description }) => {
    // Mock implementation
    return `SELECT * FROM users WHERE ${description}`;
  },
});

// Initialize the language model
const llm = new ChatOpenAI({
  modelName: "gpt-4",
  temperature: 0,
});

// Create the tools array
const tools = [queryDatabase, getTableSchema, listTables, generateSql];

// Create the SQL agent executor
const sqlAgent = new AgentExecutor({
  tools: tools,
  agent: llm,
  verbose: true,
  maxIterations: 10,
});

// Example usage
async function main() {
  const result = await sqlAgent.invoke({
    input: "Find all active users who placed orders in the last 30 days"
  });
  console.log(result);
}

if (require.main === module) {
  main();
}

export { sqlAgent };

